### PR TITLE
First major ROLLO PR with source code, tests, and one example problem 

### PR DESCRIPTION
**Pull request by gwenchee**
This PR is huge because it contains all the source code and tests for v1.0 of ROLLO. 

PR overview: 
`/docs/`: barebones documentation (I will update this to provide examples and more explanation in a future PR) 
`/examples/`: a single FHR slab ROLLO optimization problem 
`/rollo/`: ROLLO's source code 
`/tests/`: ROLLO unit test files

Seeing that this PR is so large, perhaps different people may review different parts of the PR. 

Here's a visualization and table describing ROLLO's architecture: 
Screen Shot 2021-05-19 at 11 14 40 AM
Screen Shot 2021-05-19 at 11 15 13 AM

### CODE REVIEW

**Comment by gwenchee** [comment]
@pep8speaks pep8ify

**Comment by gwenchee** [comment]
My suggestion for splitting up the review is for one person to take each numbered list. 

1) `rollo/executor.py`, `rollo/evaluation.py`, `rollo/openmc_evaluation.py`, `tests/test_executor.py`, `tests/test_evaluation.py`, `tests/test_openmc_evaluation.py`
2) `rollo/toolbox_generator.py`, `rollo/constraints.py`,  `rollo/algorithm.py`, `rollo/backend.py`, `tests/test_toolbox_generator.py`, `tests/test_constraints.py`, `tests/test_algorithm.py`,  `tests/test_backend.py`
3) `/examples/fhr-slab`, `rollo/input_validation.py`, `__main__.py`

If you're up to review my PR, comment here which number you will take. 

@munkm @smpark7 @robfairh @abachma2 @samgdotson

**Comment by samgdotson** [comment]
oh god

**Comment by samgdotson** [comment]
I'll review item 2...

**Comment by smpark7** [comment]
I'll take 1

**Comment by abachma2** [comment]
I'll take number 3

**Review by abachma2 — APPROVED** [review]
The part 3 files look good to me. Some comments to make things a little more clear and readable. Nice work!

File: README.md [hunk 1]
```diff
@@ -1,2 +1,31 @@
...
+`python -m pip install rollo`
+
+## Running ROLLO 
+`python -m rollo -i <input file> -c <checkpoint file>`
+
+Command line flags: 
+| Flag | Description | Mandatory ? |
+| ----------- | ----------- | ----------- |
+| -i | name of input file | Yes |
+| -c| name of checkpoint file | No |
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply README.md:17]: I added some extra description for it in the readme

File: examples/fhr-slab/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+**ROLLO input file and supporting OpenMC templating scripts for maximizing keff
+in a FHR fuel slab by varying the slab's TRISO fuel particle distribution.**
+
+`rollo_input.json`: ROLLO input file 
+
+`openmcinp.py`: OpenMC templating script
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply examples/fhr-slab/README.md]: done

File: rollo/input_validation.py [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+            input_algorithm, "mating_operator", {"operator": "cxOnePoint"}
+        )
+        reloaded_input_dict = input_dict.copy()
+        reloaded_input_dict["evaluators"] = input_evaluators
+        reloaded_input_dict["algorithm"] = input_algorithm
+        return reloaded_input_dict
+
+    def default_check(self, input_dict, variable, default_val):
+        """Checks if a variable is missing from a dict, and adds a default
+        value if it is
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/input_validation.py]: I changed the name of the `adds_defaults` function to `add_all_defaults` and also updated both their documentation to be more descriptive

**Comment by gwenchee** [comment]
Thanks for the review @abachma2, I have addressed your comments.

**Review by abachma2 — APPROVED** [review]
Thanks for making the changes! Looks great Gwen!

**Comment by gwenchee** [comment]
I deleted the `/docs/` directory from this branch and will make a separate PR for documentation. There are no other changes.

File: rollo/openmc_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import glob
+import openmc
+
+
+class OpenMCEvaluation:
+    """This class holds functions to evaluate openmc output files.
+    - To add new openmc outputs, create a function called "evaluate_###" and
```
Comment (smpark7) [inline rollo/openmc_evaluation.py:7]: It would be clearer if you refer to `outputs` as `output variables` or something more specific.

**Review by smpark7 — CHANGES_REQUESTED** [review]
Here's my review. Looks good and mostly clean so far

File: rollo/executor.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        Returns
+        -------
+        data : dict
+            json input file converted into a dict
+
+        """
+
+        with open(self.input_file) as json_file:
+            data = json.load(json_file)
+        return data
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/executor.py:96]: Ya, I do have that implemented in `input_validation.py` which Amanda reviewed.

[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/executor.py:96]: See lines 47-49 of `executor.py`

File: rollo/executor.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+    def organize_input_output(self, input_dict):
+        """Numbers the control variables and output variables
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/executor.py:99]: The function is actually labeling it. I updated the doc string to be more clear.

File: rollo/openmc_evaluation.py (hunk 4 above)
Comment (gwenchee) [reply rollo/openmc_evaluation.py]: I updated the docstring to be clearer.

File: rollo/executor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smpark7) [reply rollo/executor.py:96]: Okay thats great!

**Review by smpark7 — APPROVED** [review]

**Comment by gwenchee** [comment]
Thanks @smpark7!

File: tests/test_evaluation.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+import pytest
+import os
+import shutil
+from rollo.evaluation import Evaluation
+from collections import OrderedDict
+from deap import base, creator
+
+
+def test_eval_fn_generator():
```
Comment (samgdotson) [inline tests/test_evaluation.py:9]: same thing as before. What does this test do?

File: rollo/backend.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
+from deap import tools
+import pickle
+import numpy
+import time
+
+
+class BackEnd(object):
```
Comment (samgdotson) [inline rollo/backend.py:7]: The purpose of this class and the ``Algorithm`` class seems to have some overlap.

It's not totally clear what the distinction is between these two classes.

File: rollo/algorithm.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+    MPI.pickle.__init__(dill.dumps, dill.loads)
+    from mpi4py.futures import MPICommExecutor
+except:
+    warnings.warn(
+        "Failed to import mpi4py. (Only important for parallel method: mpi_evals)"
+    )
+
+
+class Algorithm(object):
```
[the remark this answers is not in the document]
Comment (samgdotson) [reply rollo/algorithm.py:18]: If this runs (some might say evaluates... or executes... ) the algorithm, what is the purpose of the ``Evaluation`` class?

File: rollo/constraints.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    def __init__(self, output_dict, input_constraints, toolbox):
+        self.constraints = self.constraints_list(input_constraints)
+        self.numbered_oup_dict = self.output_dict_numbered(output_dict)
+        self.ops = {
+            "<": operator.lt,
+            "<=": operator.le,
+            "==": operator.eq,
+            "!=": operator.ne,
+            ">=": operator.ge,
+            ">": operator.gt,
```
Comment (samgdotson) [inline rollo/constraints.py:41]: What if someone accidentally enters `` !=`` rather than ``!=``? Is there a test for this (with an errant space)?

**Review by samgdotson — COMMENTED** [review]
I haven't run the examples yet. I thought the names of some of the classes was kind of confusing. The hierarchy and the relationship between the classes is not obvious, to me. It's also not obvious what some of your tests are doing. There are a lot of integration tests and not a lot of unit tests. 
What happens if someone writes an operator in a slighty different way than is present in the dictionary? This kind of error isn't handled and could become a headache for users. 

You've obviously put a lot of thought and effort into this and I don't want to get in the way of the first big release so I'm not requesting changes. However, some of the items I mentioned could become separate issues that could be addressed later.

File: rollo/algorithm.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+try:
+    from mpi4py import MPI
+    import dill
+
+    MPI.pickle.__init__(dill.dumps, dill.loads)
+    from mpi4py.futures import MPICommExecutor
+except:
+    warnings.warn(
+        "Failed to import mpi4py. (Only important for parallel method: mpi_evals)"
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/algorithm.py:14]: I have 2 methods for parallelization that I will outline in the docs: `multiprocessing` and `mpi_evals`.  `mpi_evals` is a unique implementation that is helpful when using a supercomputer.

File: rollo/algorithm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/algorithm.py:18]: See https://github.com/arfc/rollo/pull/6#issue-647834703 to understand why I have an Executor class, an Algorithm class, and also an Evaluation class.

File: tests/test_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    selected_pop = a.apply_selection_operator(cloned_pop)
+    expected_inds = [toolbox.clone(ind) for ind in pop]
+    expected_inds.sort(key=lambda x: x[2])
+    expected_inds = expected_inds[k:]
+    for s in selected_pop:
+        assert s in expected_inds
+    os.remove("checkpoint.pkl")
+
+
+def test_apply_mating_operator():
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply tests/test_algorithm.py:186]: Every unit test is named after the function in the source code that it tests. So, I think it is clear what it is doing. For example, the function `test_apply_mating_operator` in `test_algorithm.py` tests the function `apply_mating_operator` in `algorithm.py`.

File: tests/test_evaluation.py (hunk 7 above)
Comment (gwenchee) [reply tests/test_evaluation.py:9]: See above comment.

File: rollo/backend.py (hunk 8 above)
Comment (gwenchee) [reply rollo/backend.py:7]: See #6 (comment) to understand why I have an Algorithm class, and BackEnd Class

File: rollo/constraints.py (hunk 10 above)
Comment (gwenchee) [reply rollo/constraints.py:41]: Yup the validation occurs in `input_validation.py` on line 331: https://github.com/gwenchee/rollo/blob/c56dd37fea468c9a3b61526899f1a758c1aea410/rollo/input_validation.py#L331

**Comment by gwenchee** [comment]
Thanks for the review @samgdotson. 
> I haven't run the examples yet. I thought the names of some of the classes was kind of confusing. The hierarchy and the relationship between the classes is not obvious, to me. 

An overview of the class relationships are outlined in the comment I included at the top of this PR: https://github.com/arfc/rollo/pull/6#issue-647834703

> It's also not obvious what some of your tests are doing. There are a lot of integration tests and not a lot of unit tests.

Every single test is actually a unit test. As mentioned in the above comments, every unit test is named after the function in the source code that it tests. For example, the function test_apply_mating_operator in test_algorithm.py tests the function apply_mating_operator in algorithm.py.

File: rollo/algorithm.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (samgdotson) [reply rollo/algorithm.py]: So then there are two tasks that need to be done: 
1. Add an explanation differentiating two types of parallelization in the docs.
2. Make the warning clearer. 

Maybe add "reverting to multiprocessing" or something to that effect.

[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/algorithm.py]: Yup 1. is definitely being done. Yeah I agree about 2), let me make an issue for it: https://github.com/arfc/rollo/issues/9

File: rollo/algorithm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (samgdotson) [reply rollo/algorithm.py:18]: I'm confused by the documentation, then. Reading the docstrings from each class does not give a sense for the overall organization of the ``rollo`` package.


```py
class Evaluation:    
    """
    Holds functions that generate and execute the evaluation solver's scripts.

    ..."""

class Algorithm(object):    
    """
    Builds and runs Genetic Algorithms.    
    This class holds functions to generate a generic genetic algorithm.

    ..."""

class Executor(object):    
    """
    Executes rollo simulation from start to finish.    
    Instances of this class can be used to perform a rollo run.
    ..."""
```

I'm not trying to be pedantic about this. It really isn't clear to me from reading the docstrings what each class should be used for. You have great explanations in a LaTeX table in your comment #6, those explanations should be inside the doc strings for your classes.

[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/algorithm.py:18]: Okay i hear you on that, I'll add more description for those classes.

File: tests/test_algorithm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (samgdotson) [reply tests/test_algorithm.py:186]: For which cases? Only one case? The docstring doesn't need to be long, just enough to indicate what is being tested and what case it covers. For example:

```py
def test_grid_optimize_1():    
"""    Case 1: Too many variables given    
"""    
    with pytest.raises(AssertionError):        
        loss = grid_optimizer(X_in.T,                              
        params,                              
        args=['trainlen', 'rho', 'noise'],                              
        xset=trainingLengths,                              
        verbose=False,                              
        visualize=False)
    return
```
Of course, writing tests takes time. I don't want to get in the way of progress so you may consider creating an issue for this. At the very least, you could number each test you have ``test_func_1`` and add a docstring as shown in my example s.t. future tests can be added so that it's obvious what each test does and what edge/corner cases still need to be covered.
I think it's fine to not have a complete suite of tests in this first major PR. In fact, it provides a great opportunity for undergrads to practice writing tests. This has three benefits:
1. Undergrads get experience coding
2. You get a more robust software package
3. Rollo gets an expanded user base

In sum: Number your tests, add docstrings.

File: tests/test_evaluation.py (hunk 7 above)
Comment (samgdotson) [reply tests/test_evaluation.py:9]: image

File: rollo/constraints.py (hunk 10 above)
Comment (samgdotson) [reply rollo/constraints.py:41]: Great, you have some tests for this as well. Although, my comment regarding cases and docstrings still applies.

File: rollo/algorithm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/algorithm.py:20]: I just added more descriptions. I appreciate the pedantic-ness, it helps keep me accountable with my documentation. :)

**Review by samgdotson — CHANGES_REQUESTED** [review]
Based on the strucure and design I can tell a lot of thought went into this project. It is well done. The changes I'm requesting are primarily for clarity.
1. Class docstrings should have enough information to differentiate tasks among them. 
2. Tests should have a doc string and a case number. 
3. There should be more tests. I would be satisfied with an open issue on this point.

File: tests/test_algorithm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gwenchee) [reply tests/test_algorithm.py:186]: Could you share the resource you used that suggests that tests should be numbered? I've not come across this style of test naming (I've been referring to openmc and numpy). Sure, I can make an issue to expand testing: https://github.com/arfc/rollo/issues/10

[the remark this answers is not in the document]
Comment (samgdotson) [reply tests/test_algorithm.py:186]: Looking at my own test doc string, I can see that it could be clearer also! 
"Which variables? Oh, the 'args' parameter."

**Comment by gwenchee** [comment]
Thanks for the review @samgdotson . 

1) I've addressed this with updated docstrings. 
2) As for this one, could you share the resource you used that suggests that tests should be numbered? I've not come across this style of test naming (I've been referring to openmc and numpy). 
3) See #10

File: tests/test_algorithm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (samgdotson) [reply tests/test_algorithm.py:186]: The numbering is probably just a "me" thing. Often a test will be something like: ``test_func_case`` followed by a docstring elaborating on the case. I couldn't come up with succinct enough test names so I tend to just number them.

[the remark this answers is not in the document]
Comment (gwenchee) [reply tests/test_algorithm.py:186]: Oh I see, in that case, I think i'm comfortable with my naming convention since it pairs with the function in the source file.

[the remark this answers is not in the document]
Comment (gwenchee) [reply tests/test_algorithm.py:186]: But perhaps, I'll consider expanding to numbering as I or an undergrad addresses #10

[the remark this answers is not in the document]
Comment (robfairh) [reply tests/test_algorithm.py:186]: I think Sam's suggestion is sensible. If in the future you wanted to add more tests, you just add them, without changing the names. Also, in some cases, you could have interior, edge, and cases (see Katy's book). Having that either in the name of the function, with numbers, or the docstrings would be helpful.

[the remark this answers is not in the document]
Comment (robfairh) [reply tests/test_algorithm.py:186]: It is also a good idea to keep everything consistent. If you don't have any docstring for your tests, keep it that way, because changing it at this point might be too cumbersome. Also you could have an undergrad to do that, add docstrings to all the tests.

**Comment by gwenchee** [comment]
If it's all good, can we merge? @samgdotson

**Comment by gwenchee** [comment]
Hi @samgdotson @smpark7 @abachma2 could someone merge this?
